Marlin/src/HAL/shared/Delay.h

```diff
@@ -98,42 +98,22 @@ void calibrate_delay_loop();
 
 #elif defined(__AVR__)
 
-  #define nop() __asm__ __volatile__("nop;\n\t":::)
-
-  FORCE_INLINE static void __delay_4cycles(uint8_t cy) {
-    __asm__ __volatile__(
-      L("1")
-      A("dec %[cnt]")
-      A("nop")
-      A("brne 1b")
-      : [cnt] "+r"(cy)  // output: +r means input+output
-      :                 // input:
-      : "cc"            // clobbers:
-    );
-  }
-
   // Delay in cycles
-  FORCE_INLINE static void DELAY_CYCLES(uint16_t x) {
-
-    if (__builtin_constant_p(x)) {
-      #define MAXNOPS 4
-
-      if (x <= (MAXNOPS)) {
-        switch (x) { case 4: nop(); case 3: nop(); case 2: nop(); case 1: nop(); }
-      }
-      else {
-        const uint32_t rem = (x) % (MAXNOPS);
-        switch (rem) { case 3: nop(); case 2: nop(); case 1: nop(); }
-        if ((x = (x) / (MAXNOPS)))
-          __delay_4cycles(x); // if need more then 4 nop loop is more optimal
-      }
-
-      #undef MAXNOPS
+  FORCE_INLINE static void DELAY_CYCLES(uint16_t cycles) {
+    if (__builtin_constant_p(cycles)) {
+      __builtin_avr_delay_cycles(cycles);
+    }
+    else {
+      __asm__ __volatile__(
```

Comment:
Nice! Much cleaner than my implementation :)

My only gripe would be that this effectively lowers the resolutiuon of DELAY_CYCLES. in `Delay.cpp` in dev mode, there is a call to `DELAY_CYCLES( 1);` - granted, this would be handled with `__builtin_avr_delay_cycles()`, but if there is anything else in the code that tries to delay for <4 cycles, it will still get 4 cycles of delay. It may be best to keep `__delay_4cycles()` around.

Reply:
Good catch. The intrinsic function has only a resolution of 2 cycles (which results in `DELAY_CYCLES(1)` not being delayed at all). I will try your suggestion (sometime today).